Fix compiler error

Ignition opened the pull request:
clang 17 was giving the following:

error: call to undeclared function 'rdtsc_get_tsc_hz'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]

gabrieleara commented:
It's been a while since I last used this project, but the function you mention was already defined in this header:
```
include/librdtsc/generic/cycles.h
```

Is the build error in one of your files or in one of the C files of this project?

Unless I'm missing something, I think you are including only the intermediate headers instead of the complete header `librdtsc/rdtsc.h`.

Consider that the project structure was made to reflect how the code was originally structured in DPDK sources.

Ignition commented:
```
➜  librdtsc git:(master) CC=clang-17 CXX=clang++-17 cmake -B build -S . -D CMAKE_BUILD_TYPE=Release
CMake Warning (dev) at /usr/share/cmake-3.28/Modules/GNUInstallDirs.cmake:243 (message):
  Unable to determine default CMAKE_INSTALL_LIBDIR directory because no
  target architecture is known.  Please enable at least one language before
  including GNUInstallDirs.
Call Stack (most recent call first):
  CMakeLists.txt:20 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is Clang 17.0.6
-- The CXX compiler identification is Clang 17.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang-17 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-17 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: /home/gareth/checkout/librdtsc/build
➜  librdtsc git:(master) cmake --build build -- -j 
[ 16%] Building C object CMakeFiles/rdtsc.dir/src/cycles.c.o
[ 50%] Building C object CMakeFiles/rdtsc.dir/src/timer.c.o
[ 50%] Building C object CMakeFiles/rdtsc.dir/src/common_timer.c.o
In file included from /home/gareth/checkout/librdtsc/src/cycles.c:6:
In file included from /home/gareth/checkout/librdtsc/src/x86/cycles.c:13:
/home/gareth/checkout/librdtsc/include/librdtsc/common_timer.h:21:28: error: call to undeclared function 'rdtsc_get_tsc_hz'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   21 |     register uint64_t hz = rdtsc_get_tsc_hz();
      |                            ^
/home/gareth/checkout/librdtsc/include/librdtsc/common_timer.h:21:28: note: did you mean 'rdtsc_get_tsc_freq'?
/home/gareth/checkout/librdtsc/include/librdtsc/common_timer.h:12:17: note: 'rdtsc_get_tsc_freq' declared here
   12 | extern uint64_t rdtsc_get_tsc_freq();
      |                 ^
   13 | 
   14 | #define MS_PER_S 1000
   15 | #define US_PER_S 1000000
   16 | #define NS_PER_S 1000000000
   17 | 
   18 | static inline uint64_t rdtsc_elapsed_unit(uint64_t before, uint64_t after,
   19 |                                           uint64_t multiplier) {
   20 | #define static_cast(t, x) ((t)(x))
   21 |     register uint64_t hz = rdtsc_get_tsc_hz();
      |                            ~~~~~~~~~~~~~~~~
      |                            rdtsc_get_tsc_freq
1 error generated.
gmake[2]: *** [CMakeFiles/rdtsc.dir/build.make:76: CMakeFiles/rdtsc.dir/src/cycles.c.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/rdtsc.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2

```

Ignition commented:
> ... was already defined in this header:
> 
> ```
> include/librdtsc/generic/cycles.h
> ```

That header is not included by `include/librdtsc/common_timer.h`


> Is the build error in one of your files or in one of the C files of this project?

I have no code yet, so error is from this project.

Ignition commented:
Another option is to downgrade the compiler error to a warning
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index ee9b58c..8af6b06 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,6 +28,9 @@ project(librdtsc
   DESCRIPTION "A library to read the TSC on multiple platforms."
 )
 
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "16.0")
+    add_compile_options(-Wno-error=implicit-function-declaration)
+endif()
 
 # +--------------------------------------------------------+
 # |                 Compilation variables                  |
```

https://www.redhat.com/en/blog/new-warnings-and-errors-clang-16

Ignition committed:
Cleanup code for newer compilers

Tested with:
clang-14 SUCCESS
clang-15 SUCCESS
clang-17 SUCCESS
clang-18 SUCCESS
clang-20 SUCCESS
gcc-11   SUCCESS
gcc-12   SUCCESS
gcc-13   SUCCESS
gcc-14   SUCCESS

Ignition commented:
@gabrieleara any chance of accepting these fixes for allowing newer and stricter compilers?
